sdks/python/apache_beam/transforms/util.py

```diff
@@ -780,6 +783,48 @@ def expand(self, pcoll):
             self.max_buffering_duration_secs,
             self.clock))
 
+  @typehints.with_input_types(Tuple[K, V])
+  @typehints.with_output_types(Tuple[K, Iterable[V]])
+  class WithShardedKey(PTransform):
+    """A GroupIntoBatches transform that outputs batched elements associated
+    with sharded input keys.
+
+    The sharding is determined by the runner to balance the load during the
+    execution time. By default, it spreads the input elements with the same key
+    to all available threads executing the transform.
+    """
+    def __init__(self, batch_size, max_buffering_duration_secs=None):
+      """Create a new GroupIntoBatches.WithShardedKey.
+
+      Arguments:
+        batch_size: (required) How many elements should be in a batch
+        max_buffering_duration_secs: (optional) How long in seconds at most an
+          incomplete batch of elements is allowed to be buffered in the states.
+          The duration must be a positive second duration and should be given as
+          an int or float.
+      """
+      self.batch_size = batch_size
+
+      if max_buffering_duration_secs is not None:
+        assert max_buffering_duration_secs > 0, (
+            'max buffering duration should be a positive value')
+      self.max_buffering_duration_secs = max_buffering_duration_secs
+
+    _pid = os.getpid()
```

Comment:
Why is the process id needed?

Note that there is only 1 python process per container, and since all containers start the same way it is very likely that pid for each python process in their own containers might even be the same. 

If you need to distinguish per worker, you may need some concept of worker id (which does not exist in Beam).

Reply:
I see. The pid is probably not a good choice then. I changed it to use uuid instead, analogous to the Java implementation:
https://github.com/apache/beam/blob/a016ba5632e955af25cedfbc7a9bf93c9ed858ff/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupIntoBatches.java#L96

Reply:
Yes both UUID and making it analogous to Java are good ideas. For my understanding what is the purpose of this key selection? Are we trying to by default to keep key space equal to the size of whole set of threads across all workers?

Reply:
The parallelism of the`GroupIntoBatchesDoFn` is tied to the number of keys (due to the per-key state semantics and the implementation of keyed state management). So choosing one shard per key (i.e., without key sharding) effectively means that we can not have more parallelism than the number of input keys. We are trying to by default spread the input elements to all available threads across workers, which is definitely not ideal but slightly better than no sharding.

Reply:
Sounds good. The updated comment also better explains this.